src/main/java/com/ocient/jdbc/XGConnection.java

```diff
@@ -4654,39 +4691,52 @@ private void sendClose() throws Exception
 		}
 	}
 
-	public int sendParameterMessage(final ClientWireProtocol.SetParameter param) throws SQLException
+	public CompletionStage<Integer> sendParameterMessage(final ClientWireProtocol.SetParameter param) throws SQLException
 	{
 		final ClientWireProtocol.Request.Builder builder = ClientWireProtocol.Request.newBuilder();
 		builder.setType(ClientWireProtocol.Request.RequestType.SET_PARAMETER);
 		builder.setSetParameter(param);
 		final Request wrapper = builder.build();
 
-		try
-		{
-			out.write(intToBytes(wrapper.getSerializedSize()));
-			wrapper.writeTo(out);
-			out.flush();
-			getStandardResponse();
-		} catch (final Exception ex)
+		try {
+			CompletionStage<ClientWireProtocol.ConfirmationResponse> completionStage = sendRequest(
+					wrapper,
+					wireMessage -> {
+						ClientWireProtocol.ConfirmationResponse.Builder br = ClientWireProtocol.ConfirmationResponse.newBuilder();
+						br.mergeFrom(new ByteBufInputStream(wireMessage));
+						return br.build();
+					},
+					// Reponse is already of type ConfirmationResponse
+					// identity function is sufficient
+					id -> id
+				);
+
+			// Return zero on success
+			return completionStage.thenApply(confirmationReponse -> 0);
+		}
+		catch (final Exception ex)
 		{
 			if(ex instanceof SQLException && SQLStates.SESSION_EXPIRED.equals((SQLException) ex)){
 				LOGGER.log(Level.INFO, "sendParameterMessage() received session expired. Attempting to refresh session");
-				// Refresh the now-expired session.
-				// FIXME this is now async
-				try {
-					CompletableFutures.blockingGet(refreshSession().toCompletableFuture());
-				}
-				catch(Exception e) {}
-				// Now we should be able to re-run the command.
-				return sendParameterMessage(param);
+				// Refresh the now-expired session and then
+				// re-run the command
+				return refreshSession().thenCompose(v -> {
+					try {
+						// FIXME Do we need to unroll the stack here?
+						return sendParameterMessage(param).thenApply(confirmationReponse -> 0);
+					} catch (SQLException e) {
+						LOGGER.log(Level.WARNING, String.format("Failed sending set parameter request to the server after refresh with exception %s with message %s", e, e.getMessage()));
+					}
+					return null;
+				});
 			}				
+
 			LOGGER.log(Level.WARNING, String.format("Failed sending set parameter request to the server with exception %s with message %s", ex, ex.getMessage()));
 			throw SQLStates.newGenericException(ex);
 		}
-		return 0;
 	}
 
-	private void sendSetSchema(final String schema) throws Exception
+	private CompletionStage<Void> sendSetSchema(final String schema) throws Exception
```

Comment:
Can return a `SQLResult` now and remove the exception from the method signature
```suggestion
	private CompletionStage<SQLResult<Void>> sendSetSchema(final String schema)
```